File: src/main/java/com/budiyev/android/imageloader/ImageLoader.java
/*
 * MIT License
 *
 * Copyright (c) 2017 Yuriy Budiyev [yuriy.budiyev@yandex.ru]
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
package com.budiyev.android.imageloader;

import java.util.concurrent.ExecutorService;

import android.app.Application;
import android.content.ComponentCallbacks2;
import android.content.Context;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

/**
 * Image loader is a universal tool for loading bitmaps efficiently in Android
 *
 * @see #with
 * @see #builder
 */
public final class ImageLoader {
    private final PauseLock mPauseLock = new PauseLock();
    private final ImageCache mMemoryCache;
    private final ImageCache mStorageCache;
    private final ImageRequestFactory mImageRequestFactory;

    /**
     * Image Loader
     *
     * @see #with
     * @see #builder
     */
    ImageLoader(@NonNull Context context, @NonNull ExecutorService executor, @Nullable ImageCache memoryCache,
            @Nullable ImageCache storageCache) {
        mMemoryCache = memoryCache;
        mStorageCache = storageCache;
        mImageRequestFactory = new ImageRequestFactory(context, executor, mPauseLock, memoryCache, storageCache);
    }

    /**
     * Create new load image request
     */
    @NonNull
    public ImageRequestFactory request() {
        return mImageRequestFactory;
    }

    /**
     * Create new load image request,
     * note that default data descriptor will be used with {@link ImageRequest#from}
     * and {@link ImageRequest#size} methods, source data's toString() method result will be
     * used for key generation, make sure that toString() implementation of your data class is
     * suitable for key generation, recommended to use {@link ImageRequest#descriptor} method to provide
     * unique key and some other characteristics of your custom data type instead
     *
     * @param loader Bitmap loader for specified data type
     * @return New load image request
     */
    @NonNull
    public <T> ImageRequest<T> request(@NonNull BitmapLoader<T> loader) {
        return mImageRequestFactory.custom(loader);
    }

    /**
     * Create new load image request
     *
     * @param loader  Bitmap loader
     * @param factory Data descriptor factory
     * @return New load image request
     */
    @NonNull
    public <T> ImageRequest<T> request(@NonNull BitmapLoader<T> loader, @NonNull DataDescriptorFactory<T> factory) {
        return mImageRequestFactory.custom(loader, factory);
    }

    /**
     * Delete cached image for specified {@link DataDescriptor}
     *
     * @see DataDescriptor
     */
    public void invalidate(@NonNull DataDescriptor<?> descriptor) {
        String key = descriptor.getKey();
        if (key == null) {
            return;
        }
        ImageCache memoryCache = mMemoryCache;
        if (memoryCache != null && descriptor.isMemoryCachingEnabled()) {
            memoryCache.remove(key);
        }
        ImageCache storageCache = mStorageCache;
        if (storageCache != null && descriptor.isStorageCachingEnabled()) {
            storageCache.remove(key);
        }
    }

    /**
     * Whether loading is currently paused
     */
    public boolean isLoadingPaused() {
        return mPauseLock.isPaused();
    }

    /**
     * Whether to pause image loading. If this method is invoked with {@code true} parameter,
     * all loading actions will be paused until it will be invoked with {@code false}.
     */
    public void setPauseLoading(boolean paused) {
        mPauseLock.setPaused(paused);
    }

    /**
     * Set all loading tasks to finish before any loading actions started
     */
    public void setInterruptLoadingEarly(boolean interrupt) {
        mPauseLock.setInterruptEarly(interrupt);
    }

    /**
     * Clear memory cache;
     * for better memory management when one singleton loader instance used across the app,
     * this method should be called in {@link Application#onTrimMemory)} or {@link ComponentCallbacks2#onTrimMemory},
     * default instance ({@link #with}) automatically cares about it
     *
     * @see ComponentCallbacks2
     * @see Context#registerComponentCallbacks
     */
    public void clearMemoryCache() {
        ImageCache memoryCache = mMemoryCache;
        if (memoryCache != null) {
            memoryCache.clear();
        }
    }

    /**
     * Clear storage cache
     */
    public void clearStorageCache() {
        ImageCache storageCache = mStorageCache;
        if (storageCache != null) {
            storageCache.clear();
        }
    }

    /**
     * Clear all caches
     */
    public void clearAllCaches() {
        clearMemoryCache();
        clearStorageCache();
    }

    /**
     * Get default image loader instance,
     * automatically cares about memory and storage caching
     */
    @NonNull
    public static ImageLoader with(@NonNull Context context) {
        return ImageLoaderHolder.get(context);
    }

    /**
     * Create new image loader builder instance
     */
    @NonNull
    public static ImageLoaderBuilder builder(@NonNull Context context) {
        return new ImageLoaderBuilder(context);
    }
}

File: src/main/java/com/budiyev/android/imageloader/ImageLoader.kt
/*
 * MIT License
 *
 * Copyright (c) 2018 Yuriy Budiyev [yuriy.budiyev@yandex.ru]
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
package com.budiyev.android.imageloader

import android.app.Application
import android.content.ComponentCallbacks2
import android.content.Context
import android.net.Uri
import android.os.Handler
import java.io.File
import java.io.FileDescriptor
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.ExecutorService
import kotlin.reflect.KClass

/**
 * Image loader is a universal tool for loading bitmaps efficiently in Android
 *
 * @see .with
 *
 * @see .builder
 */
class ImageLoader internal constructor(private val context: Context, private val loadExecutor: ExecutorService,
        private val cacheExecutor: ExecutorService, private val memoryCache: ImageCache?,
        private val storageCache: ImageCache?) {
    private val mainThreadHandler = Handler(context.mainLooper)
    private val pauseLock = PauseLock()
    private val descriptorFactories = ConcurrentHashMap<String, DataDescriptorFactory<Any>>()
    private val bitmapLoaders = ConcurrentHashMap<String, BitmapLoader<Any>>()

    /**
     * Whether to pause image loading. If this property is set to `true`,
     * all loading actions will be paused until it will be set `false`
     * or [isInterruptEarly] will be set to `true`
     */
    var isLoadingPaused: Boolean
        get() = pauseLock.isPaused
        set(value) {
            pauseLock.isPaused = value
        }

    /**
     * Whether all loading tasks to finish before any loading actions started
     */
    var isInterruptEarly: Boolean
        get() = pauseLock.isInterruptEarly
        set(value) {
            pauseLock.isInterruptEarly = value
        }

    /**
     * Create new image request
     * <br></br><br></br>
     * **Data types, supported by default:**
     *
     *  * [Uri], [String] - URI (remote and local)
     *  * [File] - File
     *  * [Integer] - Android resource
     *  * [FileDescriptor] - File descriptor
     *  * `byte[]` - Byte array
     *
     *
     * @param data Source data, any registered data type
     * @return New image request for specified data
     * @throws IllegalArgumentException if specified data type is not registered
     * @see .registerDataType
     */
    @Suppress("UNCHECKED_CAST")
    fun <T : Any> from(data: T): ImageRequest<T> {
        val dataClassName = data.javaClass.name
        val descriptorFactory = descriptorFactories[dataClassName] as? DataDescriptorFactory<T>
        val bitmapLoader = bitmapLoaders[dataClassName] as? BitmapLoader<T>
        if (descriptorFactory == null || bitmapLoader == null) {
            throw IllegalArgumentException("Unsupported data type: $dataClassName")
        }
        return ImageRequest(context.resources, loadExecutor, cacheExecutor, pauseLock, mainThreadHandler, memoryCache,
                storageCache, bitmapLoader, descriptorFactory.newDescriptor(data))
    }

    /**
     * Delete all cached images for specified data
     *
     * @param data Data
     * @throws IllegalArgumentException if specified data type is not registered
     * @see .registerDataType
     */
    fun invalidate(data: Any) {
        val dataClassName = data.javaClass.name
        val descriptorFactory = descriptorFactories[dataClassName] ?: throw IllegalArgumentException(
                "Unsupported data type: $dataClassName")
        InternalUtils.invalidate(memoryCache, storageCache, descriptorFactory.newDescriptor(data))
    }

    /**
     * Register data type
     *
     * @param dataClass         Source data class
     * @param descriptorFactory Data descriptor factory for specified data class
     * @param bitmapLoader      Bitmap loader factory for specified data class
     * @see DataDescriptorFactory
     *
     * @see DataDescriptor
     *
     * @see BitmapLoader
     *
     * @see .unregisterDataType
     */
    @Suppress("UNCHECKED_CAST")
    fun <T : Any> registerDataType(dataClass: Class<T>, descriptorFactory: DataDescriptorFactory<T>,
            bitmapLoader: BitmapLoader<T>) {
        val dataClassName = dataClass.name
        descriptorFactories[dataClassName] = descriptorFactory as DataDescriptorFactory<Any>
        bitmapLoaders[dataClassName] = bitmapLoader as BitmapLoader<Any>
    }

    fun <T : Any> registerDataType(dataClass: KClass<T>, descriptorFactory: DataDescriptorFactory<T>,
            bitmapLoader: BitmapLoader<T>) {
        registerDataType(dataClass.java, descriptorFactory, bitmapLoader)
    }

    /**
     * Unregister data type
     *
     * @param dataClass Source data class
     */
    fun unregisterDataType(dataClass: Class<*>) {
        val dataClassName = dataClass.name
        descriptorFactories.remove(dataClassName)
        bitmapLoaders.remove(dataClassName)
    }

    /**
     * Whether to pause image loading. If this method is invoked with `true` parameter,
     * all loading actions will be paused until it will be invoked with `false`.
     */
    fun setPauseLoading(paused: Boolean) {
        pauseLock.isPaused = paused
    }

    /**
     *
     */
    fun setInterruptLoadingEarly(interrupt: Boolean) {
        pauseLock.isInterruptEarly = interrupt
    }

    /**
     * Clear memory cache;
     * for better memory management when one singleton loader instance used across the app,
     * this method should be called in [)][Application.onTrimMemory] or [ComponentCallbacks2.onTrimMemory],
     * default instance ([.with]) automatically cares about it
     *
     * @see ComponentCallbacks2
     *
     * @see Context.registerComponentCallbacks
     */
    fun clearMemoryCache() {
        val memoryCache = memoryCache
        memoryCache?.clear()
    }

    /**
     * Clear storage cache
     */
    fun clearStorageCache() {
        val storageCache = storageCache
        storageCache?.clear()
    }

    /**
     * Clear all caches
     */
    fun clearAllCaches() {
        clearMemoryCache()
        clearStorageCache()
    }
}
